lib/ts/MemArena.h

```diff
@@ -0,0 +1,134 @@
+/** @file
+
+    Memory arena for allocations
+
+    @section license License
+
+    Licensed to the Apache Software Foundation (ASF) under one
+    or more contributor license agreements.  See the NOTICE file
+    distributed with this work for additional information
+    regarding copyright ownership.  The ASF licenses this file
+    to you under the Apache License, Version 2.0 (the
+    "License"); you may not use this file except in compliance
+    with the License.  You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+ */
+
+#ifndef _MEM_ARENA_H_
+#define _MEM_ARENA_H_
+
+#include <mutex>
+#include <ts/MemSpan.h>
+
+/// Apache Traffic Server commons.
+namespace ts
+{
+/** MemArena is a memory arena for allocations.
+
+    The intended use is for allocating many small chunks of memory - few, large allocations are best handled independently.
+    The purpose is to amortize the cost of allocation of each chunk across larger allocations in a heap style. In addition the
+    allocated memory is presumed to have similar lifetimes so that all of the memory in the arena can be de-allocatred en masse.
+
+    A generation is essentially a block of memory. The normal workflow is to freeze() the current generation, alloc() a larger and
+    newer generation, copy the contents of the previous generation to the new generation, and then thaw() the previous generation.
+    Note that coalescence must be done by the caller because MemSpan will only give a reference to the underlying memory.
+ */
+class MemArena
+{
+public:
+  /** Simple internal arena block of memory. Maintains the underlying memory.
+   */
+  struct MemArenaBlock {
+    void *data;
+    size_t size;
+    size_t allocated;
+    MemArenaBlock *next;
```

Comment:
Perhaps use `std::shared_ptr` here and one in `MemArena`. Then discarding the string of blocks is a single assignment.

Reply:
Updated to use `shared_ptr`